File: app/src/main/java/com/txt_nifty/sketch/flmml/rep/ShortBufferHolder.java
package com.txt_nifty.sketch.flmml.rep;

import android.media.AudioTrack;

public class ShortBufferHolder implements ConvertedBufferHolder {

    private final short[] buffer;

    ShortBufferHolder(int size) {
        buffer = new short[size];
    }

    @Override
    public synchronized void convertAndSet(double[] doubleBuf) {
        short[] shortBuf = buffer;
        double v;
        for (int i = 0, fin = doubleBuf.length; i < fin; i++) {
            v = doubleBuf[i];
            if (v > 1.0) v = 1.0;
            else if (v < -1.0) v = -1.0;
            shortBuf[i] = (short) (Short.MAX_VALUE * v);
        }
    }

    @Override
    public synchronized int writeTo(AudioTrack a, int s, int l) {
        return a.write(buffer, s, l);
    }
}

File: app/src/main/java/com/txt_nifty/sketch/flmml/rep/ShortBufferHolder.kt
package com.txt_nifty.sketch.flmml.rep

import android.media.AudioTrack

class ShortBufferHolder internal constructor(size: Int) : ConvertedBufferHolder {
    private val buffer: ShortArray

    init {
        buffer = ShortArray(size)
    }

    @Synchronized
    override fun convertAndSet(doubleBuf: DoubleArray) {
        val shortBuf = buffer
        for (i in doubleBuf.indices) {
            var v = doubleBuf[i]
            if (v > 1.0) v = 1.0 else if (v < -1.0) v = -1.0
            shortBuf[i] = (Short.MAX_VALUE * v).toInt().toShort()
        }
    }

    @Synchronized
    override fun writeTo(a: AudioTrack, s: Int, l: Int): Int {
        return a.write(buffer, s, l)
    }
}